### ISSUE TITLE
Vuescroll + SSR: ReferenceError: window is not defined

### ISSUE THREAD

**Issue by seconio** [YvesCoding/vuescrolljs#8]
Hello,

I have an issue with VueScroll + SSR. I use Webpack 3.12, Laravel Mix, V8JS.

According to documentation I add following lines to my main `app.js` file:

```
import vuescroll from 'vuescroll/dist/vuescroll-native';
import 'vuescroll/dist/vuescroll.css';
Vue.use(vuescroll);
```

I get:

`ReferenceError: window is not defined`

What seems to be really strange behavior is when I move the line:
`import 'vuescroll/dist/vuescroll.css';` to my `entry-client.js` file (just this one line, and yes - css) mentioned error does not occur and the page gets loaded, however when I use `<vue-scroll>` tag I get tons of following errors:

```
VM21274 entry-client.js:5621 [Vue warn]: Error in render: "TypeError: Cannot read property 'scrollPanel' of undefined"

found in

---> <VueScroll>
       <App> at resources/assets/js/components/App.vue
         <Root>
```

```
TypeError: Cannot read property '_isVue' of undefined
    at VueComponent.scrollPanelElm (VM21274 entry-client.js:96908)
    at Watcher.get (VM21274 entry-client.js:8170)
    at Watcher.evaluate (VM21274 entry-client.js:8277)
    at VueComponent.computedGetter [as scrollPanelElm] (VM21274 entry-client.js:8535)
    at VueComponent.updateNativeModeBarState (VM21274 entry-client.js:97225)
    at VueComponent.updateBarStateAndEmitEvent (VM21274 entry-client.js:97271)
    at VueComponent.<anonymous> (VM21274 entry-client.js:97255)
    at Array.<anonymous> (VM21274 entry-client.js:6865)
    at flushCallbacks (VM21274 entry-client.js:6786)
```

Those are just two random errors, but the list is long (about 20 or more). I checked the code and mentioned in errors functions are actual functions in vuescroll.js (dist folder). Installed just today with npm. Could you please give it a look? VueScroll looks awesome and I would really like to use it in my project. All SSR configuration works just fine, other plugins work without any issues. At this point it looks like the issue is not SSR, as I cannot make VueScroll work on the client-side itself as well.

**Comment by wangyi7099** [comment]
Seems not problem with SSR, see this [demo](https://github.com/YvesCoding/vuescroll-issue-list-demo). There are several questions:
1. Did you use the latest version `4.7.6` ?
2. If so, can you show me your config of your vuescroll ?
3. And the best is to provide a simple reprodution. Thanks !

**Comment by seconio** [comment]
Thank you for the reply.

1. Yes, I use latest version - in my `package.json` I have `"vuescroll": "^4.7.6"`

2., 3. First I tried the Simple example from you page `http://vuescrolljs.yvescoding.org/`, so `ops: {}`, however when I got the errors I set the following (from `http://vuescrolljs.yvescoding.org/guide/configuration.html#basic-configurations`):

```
ops: {
                    vuescroll: {
                        mode: 'native',
                        sizeStrategy: 'percent',
                        detectResize: true
                    },
                    scrollPanel: {
                        initialScrollY: false,
                        initialScrollX: false,
                        scrollingX: true,
                        scrollingY: true,
                        speed: 300,
                        easing: undefined,
                        padding: true
                    }
                }
```
Errors:

```
[Vue warn]: Error in render: "TypeError: Cannot read property 'scrollPanel' of undefined"

found in

---> <VueScroll>
       <App> at resources/assets/js/components/App.vue
         <Root>
warn @ entry-client.js:87687
logError @ entry-client.js:88831
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
Vue._render @ entry-client.js:91640
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
init @ entry-client.js:91231
hydrate @ entry-client.js:93086
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
patch @ entry-client.js:93201
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
init @ entry-client.js:91231
hydrate @ entry-client.js:93086
patch @ entry-client.js:93201
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
./resources/assets/js/entry-client.js @ entry-client.js:117132
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.867 entry-client.js:88835 TypeError: Cannot read property 'scrollPanel' of undefined
    at getPanelData (entry-client.js:100160)
    at createPanel (entry-client.js:100210)
    at Proxy.render (entry-client.js:100789)
    at VueComponent.Vue._render (entry-client.js:91638)
    at VueComponent.updateComponent (entry-client.js:89882)
    at Watcher.get (entry-client.js:90236)
    at new Watcher (entry-client.js:90225)
    at mountComponent (entry-client.js:89889)
    at VueComponent../node_modules/vue/dist/vue.esm.js.Vue.$mount (entry-client.js:95634)
    at VueComponent../node_modules/vue/dist/vue.esm.js.Vue.$mount (entry-client.js:98033)
logError @ entry-client.js:88835
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
Vue._render @ entry-client.js:91640
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
init @ entry-client.js:91231
hydrate @ entry-client.js:93086
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
patch @ entry-client.js:93201
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
init @ entry-client.js:91231
hydrate @ entry-client.js:93086
patch @ entry-client.js:93201
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
./resources/assets/js/entry-client.js @ entry-client.js:117132
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.868 entry-client.js:87687 [Vue warn]: The client-side rendered virtual DOM tree is not matching server-rendered content. This is likely caused by incorrect HTML markup, for example nesting block-level elements inside <p>, or missing <tbody>. Bailing hydration and performing full client-side render.
warn @ entry-client.js:87687
patch @ entry-client.js:93205
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
init @ entry-client.js:91231
hydrate @ entry-client.js:93086
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
hydrate @ entry-client.js:93119
patch @ entry-client.js:93201
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
init @ entry-client.js:91231
hydrate @ entry-client.js:93086
patch @ entry-client.js:93201
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
./resources/assets/js/entry-client.js @ entry-client.js:117132
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.891 entry-client.js:87687 [Vue warn]: Error in mounted hook: "TypeError: Cannot read property 'vuescroll' of undefined"

found in

---> <VueScroll>
       <App> at resources/assets/js/components/App.vue
         <Root>
warn @ entry-client.js:87687
logError @ entry-client.js:88831
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
callHook @ entry-client.js:90017
insert @ entry-client.js:91252
invokeInsertHook @ entry-client.js:93054
patch @ entry-client.js:93202
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
./resources/assets/js/entry-client.js @ entry-client.js:117132
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.891 entry-client.js:88835 TypeError: Cannot read property 'vuescroll' of undefined
    at VueComponent.setVsSize (entry-client.js:100918)
    at VueComponent.refreshInternalStatus (entry-client.js:101262)
    at VueComponent.mounted (entry-client.js:100806)
    at callHook (entry-client.js:90015)
    at Object.insert (entry-client.js:91252)
    at invokeInsertHook (entry-client.js:93054)
    at Vue.patch [as __patch__] (entry-client.js:93202)
    at Vue._update (entry-client.js:89754)
    at Vue.updateComponent (entry-client.js:89882)
    at Watcher.get (entry-client.js:90236)
logError @ entry-client.js:88835
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
callHook @ entry-client.js:90017
insert @ entry-client.js:91252
invokeInsertHook @ entry-client.js:93054
patch @ entry-client.js:93202
Vue._update @ entry-client.js:89754
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
Watcher @ entry-client.js:90225
mountComponent @ entry-client.js:89889
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:95634
./node_modules/vue/dist/vue.esm.js.Vue.$mount @ entry-client.js:98033
./resources/assets/js/entry-client.js @ entry-client.js:117132
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.947 entry-client.js:87687 [Vue warn]: Error in nextTick: "TypeError: Cannot read property '_isVue' of undefined"

found in

---> <VueScroll>
       <App> at resources/assets/js/components/App.vue
         <Root>
warn @ entry-client.js:87687
logError @ entry-client.js:88831
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
(anonymous) @ entry-client.js:88933
flushCallbacks @ entry-client.js:88852
Promise.then (async)
microTimerFunc @ entry-client.js:88900
nextTick @ entry-client.js:88944
Vue.$nextTick @ entry-client.js:91611
(anonymous) @ entry-client.js:86450
step @ entry-client.js:86178
step @ entry-client.js:86185
runQueue @ entry-client.js:86189
(anonymous) @ entry-client.js:86443
step @ entry-client.js:86178
(anonymous) @ entry-client.js:86182
(anonymous) @ entry-client.js:86428
(anonymous) @ entry-client.js:86256
iterator @ entry-client.js:86407
step @ entry-client.js:86181
step @ entry-client.js:86185
runQueue @ entry-client.js:86189
confirmTransition @ entry-client.js:86436
transitionTo @ entry-client.js:86338
init @ entry-client.js:86958
beforeCreate @ entry-client.js:85004
callHook @ entry-client.js:90015
Vue._init @ entry-client.js:91720
Vue @ entry-client.js:91823
./resources/assets/js/app.js @ entry-client.js:114195
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
./resources/assets/js/entry-client.js @ entry-client.js:117096
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.953 entry-client.js:88835 TypeError: Cannot read property '_isVue' of undefined
    at VueComponent.scrollPanelElm (entry-client.js:100834)
    at Watcher.get (entry-client.js:90236)
    at Watcher.evaluate (entry-client.js:90343)
    at VueComponent.computedGetter [as scrollPanelElm] (entry-client.js:90601)
    at VueComponent.updateNativeModeBarState (entry-client.js:101151)
    at VueComponent.updateBarStateAndEmitEvent (entry-client.js:101197)
    at VueComponent.<anonymous> (entry-client.js:101181)
    at Array.<anonymous> (entry-client.js:88931)
    at flushCallbacks (entry-client.js:88852)
logError @ entry-client.js:88835
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
(anonymous) @ entry-client.js:88933
flushCallbacks @ entry-client.js:88852
Promise.then (async)
microTimerFunc @ entry-client.js:88900
nextTick @ entry-client.js:88944
Vue.$nextTick @ entry-client.js:91611
(anonymous) @ entry-client.js:86450
step @ entry-client.js:86178
step @ entry-client.js:86185
runQueue @ entry-client.js:86189
(anonymous) @ entry-client.js:86443
step @ entry-client.js:86178
(anonymous) @ entry-client.js:86182
(anonymous) @ entry-client.js:86428
(anonymous) @ entry-client.js:86256
iterator @ entry-client.js:86407
step @ entry-client.js:86181
step @ entry-client.js:86185
runQueue @ entry-client.js:86189
confirmTransition @ entry-client.js:86436
transitionTo @ entry-client.js:86338
init @ entry-client.js:86958
beforeCreate @ entry-client.js:85004
callHook @ entry-client.js:90015
Vue._init @ entry-client.js:91720
Vue @ entry-client.js:91823
./resources/assets/js/app.js @ entry-client.js:114195
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
./resources/assets/js/entry-client.js @ entry-client.js:117096
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.968 entry-client.js:87687 [Vue warn]: Error in render: "TypeError: Cannot read property 'scrollPanel' of undefined"

found in

---> <VueScroll>
       <App> at resources/assets/js/components/App.vue
         <Root>
warn @ entry-client.js:87687
logError @ entry-client.js:88831
globalHandleError @ entry-client.js:88826
handleError @ entry-client.js:88815
Vue._render @ entry-client.js:91640
updateComponent @ entry-client.js:89882
get @ entry-client.js:90236
run @ entry-client.js:90313
flushSchedulerQueue @ entry-client.js:90075
(anonymous) @ entry-client.js:88931
flushCallbacks @ entry-client.js:88852
Promise.then (async)
microTimerFunc @ entry-client.js:88900
nextTick @ entry-client.js:88944
queueWatcher @ entry-client.js:90162
update @ entry-client.js:90303
notify @ entry-client.js:87793
reactiveSetter @ entry-client.js:88110
proxySetter @ entry-client.js:90394
(anonymous) @ entry-client.js:2824
(anonymous) @ entry-client.js:88931
flushCallbacks @ entry-client.js:88852
Promise.then (async)
microTimerFunc @ entry-client.js:88900
nextTick @ entry-client.js:88944
Vue.$nextTick @ entry-client.js:91611
(anonymous) @ entry-client.js:86450
step @ entry-client.js:86178
step @ entry-client.js:86185
runQueue @ entry-client.js:86189
(anonymous) @ entry-client.js:86443
step @ entry-client.js:86178
(anonymous) @ entry-client.js:86182
(anonymous) @ entry-client.js:86428
(anonymous) @ entry-client.js:86256
iterator @ entry-client.js:86407
step @ entry-client.js:86181
step @ entry-client.js:86185
runQueue @ entry-client.js:86189
confirmTransition @ entry-client.js:86436
transitionTo @ entry-client.js:86338
init @ entry-client.js:86958
beforeCreate @ entry-client.js:85004
callHook @ entry-client.js:90015
Vue._init @ entry-client.js:91720
Vue @ entry-client.js:91823
./resources/assets/js/app.js @ entry-client.js:114195
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
./resources/assets/js/entry-client.js @ entry-client.js:117096
__webpack_require__ @ entry-client.js:679
fn @ entry-client.js:89
1 @ entry-client.js:119034
__webpack_require__ @ entry-client.js:679
./node_modules/after/index.js.module.exports @ entry-client.js:725
(anonymous) @ entry-client.js:728
11:03:26.970 entry-client.js:88835 TypeError: Cannot read property 'scrollPanel' of undefined
    at getPanelData (entry-client.js:100160)
    at createPanel (entry-client.js:100210)
    at Proxy.render (entry-client.js:100789)
    at VueComponent.Vue._render (entry-client.js:91638)
    at VueComponent.updateComponent (entry-client.js:89882)
    at Watcher.get (entry-client.js:90236)
    at Watcher.run (entry-client.js:90313)
    at flushSchedulerQueue (entry-client.js:90075)
    at Array.<anonymous> (entry-client.js:88931)
    at flushCallbacks (entry-client.js:88852)
```
I imported VueScroll on `entry-client.js` only, SSR is off for now (bailing hydration anyway).

**Comment by wangyi7099** [comment]
Thanks for your error detail, can you give me this error code  and let me see what went wrong.

**Comment by wangyi7099** [comment]
Hi I'll check this on V8JS PHP this weekend.

**Comment by seconio** [comment]
Great, can't wait! If you have problem with setting up V8JS hit me up over mail: accounts (at) secon (dot) io, I will help you. From my experience configuration was pain in the ass

**Comment by wangyi7099** [comment]
OK!!Thanks! :)

**Comment by wangyi7099** [comment]
@seconio 
Hi seems work for me , please checkout this [repo](https://github.com/YvesCoding/vuescroll-ssr-php). You can find the differences between yours and mime. Feel free to comment if you have any questions!

**Comment by wangyi7099** [comment]
This should be fixed in `4.8.4`
Cc @seconio
